Sources/SwiftProtobuf/JSONEncoder.swift

```diff
@@ -148,13 +173,19 @@ internal struct JSONEncoder {
                 append(staticText: "\"Infinity\"")
             }
         } else {
-            // TODO: Be smarter here about choosing significant digits
-            // See: protoc source has C++ code for this with interesting ideas
             if let v = Int64(safely: value) {
                 appendInt(value: v)
             } else {
                 let s = String(value)
-                append(text: s)
+                let reparsed = Double(s)
+                // If exact match, then default precision is sufficient:
+                if value == reparsed {
+                    append(text: s)
+                } else {
+                    let precision = DBL_DIG + 2
+                    let s = String(format: "%.*g", precision, value)
```

Comment:
Re float - maybe because you are testing on macOS and in 64 bit so float is a larger data type?  Try the iOS 32bit build where float would be smaller?